tests_e2e/tests/lib/cgroup_helpers.py

```diff
@@ -0,0 +1,145 @@
+import logging
+import os
+import re
+import subprocess
+import sys
+
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.osutil import systemd
+from azurelinuxagent.common.version import DISTRO_NAME, DISTRO_VERSION
+from tests_e2e.tests.lib.agent_log import AgentLog
+
+BASE_CGROUP = '/sys/fs/cgroup'
+AGENT_CGROUP_NAME = 'WALinuxAgent'
+AGENT_SERVICE_NAME = systemd.get_agent_unit_name()
+AGENT_CONTROLLERS = ['cpu', 'memory']
+EXT_CONTROLLERS = ['cpu', 'memory']
+
+CGROUP_TRACKED_PATTERN = re.compile(r'Started tracking cgroup ([^\s]+)\s+\[(?P<path>[^\s]+)\]')
+
+GATESTEXT_FULL_NAME = "Microsoft.Azure.Extensions.Edp.GATestExtGo"
+GATESTEXT_SERVICE = "gatestext.service"
+AZUREMONITOREXT_FULL_NAME = "Microsoft.Azure.Monitor.AzureMonitorLinuxAgent"
+AZUREMONITORAGENT_SERVICE = "azuremonitoragent.service"
+MDSD_SERVICE = "mdsd.service"
+
+
+def exit_if_cgroups_not_supported():
+    logging.info("===== Checking if distro supports cgroups =====")
+
+    base_fs_exists = os.path.exists(BASE_CGROUP)
+
+    if not base_fs_exists:
+        logging.warning("\tDistro %s-%s does not support cgroups -- exiting", DISTRO_NAME, DISTRO_VERSION)
+        sys.exit(1)
+    else:
+        logging.info('\tDistro %s-%s supports cgroups\n', DISTRO_NAME, DISTRO_VERSION)
+
+
+def run_get_output(cmd, print_std_out=False):
+    # Returns a list of stdout lines without \n at the end of the line.
+    output = subprocess.check_output(cmd,
+                                     stderr=subprocess.STDOUT,
+                                     shell=True)
+    output = ustr(output,
+                  encoding='utf-8',
+                  errors="backslashreplace")
+
+    if print_std_out:
+        logging.info(output)
+
+    return output.split("\n")
+
+
+def print_cgroups():
+    logging.info("====== Currently mounted cgroups ======")
+    for m in run_get_output('mount'):
+        if 'type cgroup' in m:
+            logging.info('\t%s', m)
+    logging.info("")
+
+
+def print_processes():
+    logging.info("====== Currently running processes ======")
+    processes = run_get_output("ps aux --forest")
```

Comment:
what does '-- forest' do here

Reply:
Forest fetches child process of parent in tree fashion